Optimization Path Plotter

PaulTalbot-INL opened the pull request:
--------
Pull Request Description
--------
##### What issue does this change request address? (Use "#" before the issue to link it, i.e., #42.)
Closes #1506

##### What are the significant changes in functionality due to this change request?
Adds a `OutStreams.Plot` `subType` for plotting optimization results.

----------------
For Change Control Board: Change Request Review
----------------
The following review must be completed by an authorized member of the Change Control Board.
- [x] 1. Review all computer code.
- [x] 2. If any changes occur to the input syntax, there must be an accompanying change to the user manual and xsd schema. If the input syntax change deprecates existing input files, a conversion script needs to be added (see Conversion Scripts).
- [x] 3. Make sure the Python code and commenting standards are respected (camelBack, etc.) - See on the [wiki](https://github.com/idaholab/raven/wiki/RAVEN-Code-Standards#python) for details.
- [x] 4. Automated Tests should pass, including run_tests, pylint, manual building and xsd tests. If there are changes to Simulation.py or JobHandler.py the qsub tests must pass.
- [x] 5. If significant functionality is added, there must  be tests added to check this. Tests should cover all possible options.  Multiple short tests are preferred over one large test. If new development on the internal JobHandler parallel system is performed, a cluster test must be added setting, in <RunInfo> XML block, the node ```<internalParallel>``` to True.
- [x] 6. If the change modifies or adds a requirement or a requirement based test case, the Change Control Board's Chair or designee also needs to approve the change.  The requirements and the requirements test shall be in sync.
- [x] 7. The merge request must reference an issue.  If the issue is closed, the issue close checklist shall be done.
- [x] 8. If an analytic test is changed/added is the the analytic documentation updated/added?
- [x] 9. If any test used as a basis for documentation examples (currently found in `raven/tests/framework/user_guide` and `raven/docs/workshop`) have been changed, the associated documentation must be reviewed and assured the text matches the example.

alfoa commented on doc/user_manual/OutStreamSystem.tex:
```diff
@@ -1714,3 +1714,53 @@ \subsubsection{SamplePlot}
+
+\subsubsection{OptPath}
+The \xmlString{OptPath} \xmlAttr{subType} shows the path taken during a \xmlNode{MultiRun} using
+a gradient-based \xmlNode{Optimizer} as the sampling strategy. It is generally used to show the progressive
```
only for the gradient-based optimizers? What about the others?

PaulTalbot-INL replied:
The heuristic optimizers do not provide the same meta variables as gradient descent, since they don't work the same.

It's possible this could be extended in the future to capture the outputs of heuristic optimizations, as well; alternatively, a plotting system designed to work well for those algorithms could be implemented.

Jimmy-INL replied:
This works fine for GA. I just suggest having a different color for 'first' and 'final' and add them to the legend.

PaulTalbot-INL replied:
done

Jimmy-INL commented on framework/OutStreams/PlotInterfaces/OptPath.py:
```diff
@@ -0,0 +1,137 @@
+    self.source = None      # reference to DataObject source
+    self.sourceName = None  # name of DataObject source
+    self.vars = None        # variables to plot
+    self.markerMap = {'accepted': 'go',
```
I'd add 'first': 'yo',
             'final': 'mo'
you might pick different colors if you like :)

PaulTalbot-INL replied:
I thought about that; since two people did, it must be worth doing. I added it as you recommended.

Jimmy-INL replied:
Great, thanks, Paul. By the way, I have tried it on GA and it is working fine, it certainly can have some extra features like a degrading alpha with the population number to separate populations from each other, but I think I can do that later. The only thing that I think will be really beneficial if you can add a 2D contour plot, this also works with all optimizers and is certainly needed to explore the obj function or even the fitness in my case. so another subplot that combines every two variables. If you can add that it will be awesome.

PaulTalbot-INL replied:
I did think about this a bit, but I think this is a fairly complicated addition. The user is going to have to specify how they want to combine variables, and then we don't have an actual contour to plot since we have a rather sparse sampling of the space itself.

I recommend this be considered as a different plotter, or different PR extending this plotter. I don't have the program bandwidth to tackle that second type of plot right now.

Jimmy-INL replied:
I meant to do all the combination 2 variables at a time. About the sparsity, in most problems, if it will continue till convergence, it is most likely it won't be that sparse, and if the user wants a denser plot it is intuitive to increase the number of iterations. But I understand the bandwidth issue. I have sent two tests one for GA and one for SA via email, please add those and check them, then remove from .tex that it only works for GD, then this is ready to be merged at least for this single plot. In the future, I wish if we can add the convergence mp4s, they were really the most descriptive.

PaulTalbot-INL replied:
I completely agree, the animations are very interesting to see, and I would love to have the opportunity to automate them in some way.

Perhaps GA is different, but in GD we sample a very small portion of the overall space, making the contour plots very sparse for most problems, especially with single trajectory. For all the animations I've shown of the optimizer previously, a heavy grid sampling was performed a priori which allowed the background contour plot to be made. There may be some options there. At any rate, we can pick that up in discussion for future work.

I look forward to seeing your examples, and will add them in as tests, and modify the docs.

Jimmy-INL replied:
Thanks, Paul. Let me know once you are done. Should I merge? or do you need two reviewers?

Jimmy-INL replied:
I was also trying to make the plot for objVar ('ans') be a semiology but it was not straightforward as the self._objectiveVar is not accessible there. This is because GA and Simulated annealing sometimes end up having very slight fluctuations that are hard to see on the regular y-axis.

PaulTalbot-INL replied:
I have a vision for this, where somehow we zoom in (or semilog) somehow as a function of convergence on the solution. As you noted, though, it would take additional information from the user, or perhaps we could do a density clustering on distance from the final result for each variable, or something. Maybe a local variance as a distance to the solution? All great possible additions for the future; this isn't written in final form obviously.

I don't think I need two reviewers, and no design-bases changes are made here so your review should be plenty. I will note the tests cannot currently run due to the HPC being down, however, so I imagine we will have to wait for those.

Jimmy-INL replied:
Why don't you add an issue describing this vision as a feature request maybe one day we can do it. Let me know when the tests are ready so that I can merge them.

PaulTalbot-INL replied:
done

alfoa commented on tests/framework/Optimizers/GeneticAlgorithms/testGAMaxwRep.xml:
```diff
@@ -80,7 +81,7 @@
 
       <convergence>
-        <objective>-1</objective>
+        <AHDp>0.1</AHDp>
```
This is not in the manual. @Jimmy-INL

PaulTalbot-INL replied:
Can this be removed from the test change, or should we put off the GA demonstration for another PR?

Jimmy-INL replied:
on it

Jimmy-INL replied:
can you please convert it back to <objective>-1</objective>? If the test is long reduce the <limit>

Jimmy-INL replied:
This shouldn't affect the plotter at all. Sorry, I thought it is hanging because other tests are failing not because of this. My bad, I think I am losing my memory

PaulTalbot-INL commented:
I fixed the generation script and added the updated docs. It looks like quite a few options were hiding in the InputSpecs that hadn't made it into the docs.

This is one more big sign that we should try to automate all the docs, I suppose.

PaulTalbot-INL commented:
@Jimmy-INL this is passing tests still; I merged in devel to resolve some conflicts. Where are we at with this PR?

moosebuild commented:
Job [Test OpenSUSE Leap 15](https://civet.inl.gov/job/744327/) on 7851b4f : invalidated by @Jimmy-INL

Jimmy-INL commented:
The openSuSE Leap 15 machine is currently down. I am merging this after discussion with @PaulTalbot-INL
